# Task - 4.6: Combining & Exporting Data
**Questions 3–9**


# Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# Turn your project folder path into a string

path = r'/Users/peterreadman/Desktop/Python Projects/CareerFoundry/04-2020 Instacart Basket Analysis/'

# Question 3
In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved.

In [8]:
# import orders_products_combined dataframe from the pkl file
df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data', '02b Prepared Data', 'orders_products_combined.pkl'))

# Question 4
Check the shape of the imported dataframe (it should be the same as the one you exported—always check!)

In [13]:
df_orders_products_combined.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2539329        1             1                       2                  8   
2   2539329        1             1                       2                  8   
3   2539329        1             1                       2                  8   
4   2539329        1             1                       2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  
0                    NaN         196                  1          0   both  
1                    NaN       14084                  2          0   both  
2                    NaN       12427                  3          0   both  
3                    NaN       26088                  4          0   both  
4                    NaN       26405                  5          0   both

In [9]:
# 04. Check the shape of the imported dataframe (it should be the same as the one you exported—always check!)
df_orders_products_combined.shape

(32434489, 10)

The shape is as expected

# Question 5
Determine a suitable way to combine the orders_products_combined dataframe with your products data set. Make sure you’re using your wrangled, cleaned, and deduped products data set stored in your “Prepared Data” folder from the previous Exercise’s task.

In [16]:
# import products_checked.csv first
df_products_checked = pd.read_csv(os.path.join(path, '02 Data', '02b Prepared Data', 'products_checked.csv'))

In [15]:
# check head of products_checked df to check it imported
df_products_checked.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [17]:
# drop '_merge' column from df_orders_products_combined
df_orders_products_combined.drop('_merge', axis=1, inplace=True)

In [18]:
# check '_merge' column is dropped
df_orders_products_combined.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2539329        1             1                       2                  8   
2   2539329        1             1                       2                  8   
3   2539329        1             1                       2                  8   
4   2539329        1             1                       2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  
0                    NaN         196                  1          0  
1                    NaN       14084                  2          0  
2                    NaN       12427                  3          0  
3                    NaN       26088                  4          0  
4                    NaN       26405                  5          0

In [19]:
# join products_checked onto orders_products_combined using 'product_id' as the key
# merge on order_id 

df_merged_orders_products = df_orders_products_combined.merge(df_products_checked, on = 'product_id', indicator = True)

In [20]:
# check head of new merged dataframe
df_merged_orders_products.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  
0         195         Soda        77              7     9.0   both  
1         195         Soda        77              7     9.0   both  
2         195         Soda        77              7     9.0   both  
3         195         Soda        77              7     9.0   both  
4         195         Soda        77              7     9.0   both

# Question 6
Confirm the results of the merge using the merge flag.

In [21]:
# use the value_counts() function to sum all the values in the '_merge' column
df_merged_orders_products['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

There are 32,404,859 rows in the new merged data set, the same as the df_orders_products_combined dataset, as expected. 

# Question 7
Export the newly created dataframe as orders_products_merged in a suitable format (taking into consideration the size).

In [22]:
# export df_merged_orders_products as a pickle file due to the file size
df_merged_orders_products.to_pickle(os.path.join(path, '02 Data', '02b Prepared Data', 'orders_products_merged.pkl'))